wcfsetup/install/files/lib/acp/form/MultifactorAuthenticationForm.class.php

```diff
@@ -0,0 +1,180 @@
+<?php
+namespace wcf\acp\form;
+use wcf\data\object\type\ObjectType;
+use wcf\data\user\User;
+use wcf\form\AbstractForm;
+use wcf\form\AbstractFormBuilderForm;
+use wcf\system\application\ApplicationHandler;
+use wcf\system\cache\runtime\UserProfileRuntimeCache;
+use wcf\system\exception\IllegalLinkException;
+use wcf\system\exception\PermissionDeniedException;
+use wcf\system\request\LinkHandler;
+use wcf\system\user\multifactor\IMultifactorMethod;
+use wcf\system\user\multifactor\Setup;
+use wcf\system\WCF;
+use wcf\util\HeaderUtil;
+
+/**
+ * Represents the multi-factor authentication form.
+ *
+ * @author	Tim Duesterhus
+ * @copyright	2001-2020 WoltLab GmbH
+ * @license	GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package	WoltLabSuite\Core\Acp\Form
+ * @since	5.4
+ */
+class MultifactorAuthenticationForm extends AbstractFormBuilderForm {
```

Comment:
I really want to avoid class inheritance across frontend and ACP. During development of MFA I realized that the fact that the frontend login inherits from the ACP login causes ugly control flow.

OTOH duplicated code is not nice either. Should I look into moving the methods into a trait?

Reply:
Deriving from the frontend class (and vice versa) isn't a bad thing. Traits always feel like one more layer of indirection and drastically reduce readability in terms of code flow.

Just derive the forms from each other, none of the solutions are fine, but that one is the least awkward. Especially security releated code should not be duplicated, because it _will_ get out of sync.